src/services/polkit/gobjectref.hpp

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <glib-object.h>
+
+namespace qs::service::polkit {
+
+struct GObjectNoRefTag {};
+constexpr GObjectNoRefTag G_OBJECT_NO_REF;
+
+template <typename T>
+class GObjectRef {
+public:
+	explicit GObjectRef(T* ptr = nullptr): ptr(ptr) {
+		if (this->ptr) {
+			g_object_ref(this->ptr);
+		}
+	}
+
+	explicit GObjectRef(T* ptr, GObjectNoRefTag /*tag*/): ptr(ptr) {}
+
+	~GObjectRef() {
+		if (this->ptr) {
+			g_object_unref(this->ptr);
+		}
+	}
+
+	// We do handle self-assignment in a more general case by checking the
+	// included pointers rather than the wrapper objects themselves.
```

Comment:
The comment says this which makes sense but it looks like we dont ever compare the gobjectref content pointers before unreffing?

Reply:
Whoops yeah. My misread.